components/server/src/api/auth-provider-service-api.ts

```diff
@@ -0,0 +1,189 @@
+
+        if (!organizationId && !ownerId) {
+            throw new ConnectError("organizationId or ownerId is required", Code.InvalidArgument);
+        }
+
+        const authProviders = organizationId
+            ? await this.authProviderService.getAuthProvidersOfOrg(context.user.id, organizationId)
+            : await this.authProviderService.getAuthProvidersOfUser(context.user.id);
+
+        const redacted = authProviders.map(AuthProviderEntry.redact.bind(AuthProviderEntry));
+
+        const result = new ListAuthProvidersResponse({
+            authProviders: redacted.map((ap) => this.apiConverter.toAuthProvider(ap)),
+        });
+        return result;
+    }
+
+    /**
+     * Listing descriptions of auth providers doesn't require authentication.
+     */
+    @Unauthenticated()
+    async listAuthProviderDescriptions(
+        _request: ListAuthProviderDescriptionsRequest,
+        context: HandlerContext,
+    ): Promise<ListAuthProviderDescriptionsResponse> {
+        const user = context.user;
+        const aps = user
+            ? await this.authProviderService.getAuthProviderDescriptions(user)
+            : await this.authProviderService.getAuthProviderDescriptionsUnauthenticated();
+
+        return new ListAuthProviderDescriptionsResponse({
+            descriptions: aps.map((ap: AuthProviderInfo) => this.apiConverter.toAuthProviderDescription(ap)),
+        });
+    }
+
+    async updateAuthProvider(
+        request: UpdateAuthProviderRequest,
+        context: HandlerContext,
+    ): Promise<UpdateAuthProviderResponse> {
+        if (!request.authProviderId) {
+            throw new ConnectError("authProviderId is required", Code.InvalidArgument);
+        }
+        const clientId = request.clientId;
+        const clientSecret = request.clientSecret;
+        if (!clientId || typeof clientSecret === "undefined") {
+            throw new ConnectError("clientId or clientSecret are required", Code.InvalidArgument);
+        }
+
+        const authProvider = await this.authProviderService.getAuthProvider(context.user.id, request.authProviderId);
+        if (!authProvider) {
+            throw new ConnectError("Provider not found.", Code.NotFound);
+        }
+
+        if (authProvider.organizationId) {
+            await this.authProviderService.updateOrgAuthProvider(context.user.id, {
+                id: request.authProviderId,
+                organizationId: authProvider.organizationId,
+                clientId: clientId,
+                clientSecret: clientSecret,
+            });
+        } else {
+            await this.authProviderService.updateAuthProviderOfUser(context.user.id, {
+                id: request.authProviderId,
+                ownerId: context.user.id,
+                clientId: clientId,
+                clientSecret: clientSecret,
+            });
+        }
+
+        return new UpdateAuthProviderResponse();
```

Comment:
udpates should reuturn fully updated value, see https://www.notion.so/gitpod/Public-API-guideline-a010fa3d494544b78f28945e53f730ac?pvs=4#8360748f176948e1a4483254d8a3b12f